doc(openid-connect) how to configure the OpenID-Connect plugin with Curity (#2809)
Kong/docs.konghq.com: app/enterprise/2.3.x/plugins/oidc-curity.md

```bash
curl -X POST http://kong:8001/services/httpbin/plugins \
--data name="openid-connect" \
--data config.issuer="https://idsvr.example.com/oauth/v2/oauth-anonymous" \
--data config.client_id="gateway-client" \
--data config.client_secret="Password1" \
--data config.scopes_required="openid" \
--data config.hide_credentials="true" \
--data config.upstream_headers_claims="phantom_token" \
--data config.upstream_headers_names="phantom_token" \
--data config.auth_methods="introspection"
```

Parameter | Description | Example | Required for integration
--- |--- |--- |---
`config.issuer` | Used for discovery. Kong appends `/.well-known/openid-configuration`. Should be set to the `realm` or `iss` if no discovery endpoint is available. | `https://idsvr.example.com/oauth/v2/oauth-anonymous` | Yes
`config.client_id` | The ID of a client with the introspection capability | `gateway-client`| Yes
`config.client_secret` | Secret of the client used for introspection| `Password1` | Yes

> `config.scopes_required`| Optional scopes required in introspection result for authorization. | `openid email records_read`  | No

Reviewer: `scopes_claim` is set to `"scopes"` by default. Adding `scopes_required` turns on the extra coarse grained authorization logic. Here it means that introspection json must have a claim `scopes` that has value `openid` (among the others). It can be space separated or an json array.

Author: Added clarification

> `config.hide_credentials` | Boolean value. This will prevent the incoming Access Token from being forwarded to the upstream API. |`true` | No

Reviewer: You might need to set `config.upstream_access_token_header=` (aka nil) too because that might add it back. As this client can receive tokens with multiple names and ways, the hide_credential just removes that, but `config.upstream_access_token_header` is set by default to `authorization:bearer`, so that in this case adds it back.

Author: Added clarification

Call the exposed service created earlier and pass the opaque Access Token in the `Authorization` header.

```bash
curl -X GET http://kong:8000/httpbin/get \
--header "Authorization: Bearer <OPAQUE ACCESS TOKEN"
```

Kong will introspect the opaque token and receive the JWT in the response. The JWT is forwarded to the upstream API. Since the configured upstream API is Httpbin it will be echoed back. The below sample response shows the `phantom_token` containing the JWT and can be consumed by the API. The response is truncated for readability.

```json
{
    "args": {},
    "headers": {
        ...
        "Host": "httpbin.org",
        "Phantom-Token": "eyJraWQiOiIxN...",
        "X-Forwarded-Host": "localhost",
        "X-Forwarded-Path": "/httpbin/get",
        "X-Forwarded-Prefix": "/httpbin"
    },
    "origin": "172.27.0.1, 69.181.2.136",
    "url": "http://localhost/get"
}
```

### Resources
* Overview of the [Phantom Token Pattern][curity-phantom-token-pattern]
* Information on the [Introspection and Phantom Tokens][curity-phantom-token-introspection] flow
* Installing the [Curity Identity Server][curity-getting-started]

----

## Kong Dev Portal Authentication

> The [Kong Developer Portal][kong-dev-portal-doc] is a part of Kong Enterprise Gateway. The portal enables developers to access and manage configured services and documentation to provide a streamlined onboarding process. The Kong Dev Portal supports various options for user authentication and access control. This artiguide outlines how to configure the Kong OpenID Connect Plugin to leverage the Curity Identity Server as a third-party Identity Provider for user Authentication to the Dev Portal.

Reviewer 2: ```suggestion
The [Kong Developer Portal][kong-dev-portal-doc] is a part of {{site.ee_product_name}}. The portal enables developers to access and manage configured services and documentation to provide a streamlined onboarding process. The Kong Dev Portal supports various options for user authentication and access control. This guide outlines how to configure the Kong OpenID Connect Plugin to leverage the Curity Identity Server as a third-party Identity Provider for user Authentication to the Dev Portal.
```

### Curity Setup

The Kong Dev Portal needs a client configured in the Curity Identity Server. The [Curity Getting Started Guide][curity-getting-started] outlines details on how to configure a client. The configuration details below should be sufficient:

- Authentication method: `secret`. Make note of the Secret and use it in the `client_secret` field in the Developer Portal configuration (see below).
- Capabilities: `Code Flow`.
- Following scopes: `openid`, `profile`, and `email`.
- Choose suitable authentication methods.
- Add redirect URI (by default `http://kong:8004/<WORKSPACE_NAME>/auth`).

> NOTE: There is an issue with short-lived Access Tokens that is under investigation. Increase the `Access Token Time to Live` in the client configuration to `3000` as a temporary workaround. 

Reviewer 2: ```suggestion
NOTE: There is an issue with short-lived access tokens that is under investigation. Increase the `Access Token Time to Live` in the client configuration to `3000` as a temporary workaround. 
```

This is a Curity issue? Or Kong? Any tracking number?

#### Example Configuration Object

Below is an example configuration object that is used to configure the OIDC plugin for the Dev Portal.

```json
{
    "redirect_uri": ["https://kong-dev-portal:8004/default/auth"],
    "consumer_by": ["username","custom_id","id"],
    "leeway": 1000,
    "scopes": ["openid","profile","email"],
    "logout_query_arg": "logout",
    "login_redirect_uri": ["https://kong-dev-portal:8003"],
    "login_action": "redirect",
    "logout_redirect_uri": ["https://kong-dev-portal:8003"],
    "ssl_verify": false,
    "client_id": ["kong-dev-portal-client"],
    "forbidden_redirect_uri": ["https://kong-dev-portal:8003/unauthorized"],
    "client_secret": ["Pa$$w0rd!"],
    "issuer": "https://idsvr.example.com/oauth/v2/oauth-anonymous/",
    "logout_methods": ["GET"],
    "consumer_claim": ["email"],
    "login_redirect_mode": "query"
}
```

![Enable OIDC in Kong Dev Portal](/assets/images/docs/dev-portal/curity/kong-dev-portal.png)

### Curity Authentication Action

> An Authentication Action to automatically provision the user to the Kong Developer Portal is available in the Curity GitHub repository. Using the Action is not mandatory as the user could be provisioned in other ways, such as manually through the Kong Developer portal login screen. However, using the Authentication Action would streamline the user flow as the Action will take the user's full name and the email from the Curity Authenticator and automatically provision that to the Kong Developer Portal using the exposed API.

Reviewer 2: ```suggestion
An Authentication Action to automatically provision the user to the Kong Developer Portal is available in the Curity GitHub repository. Using the Action is not mandatory as the user could be provisioned in other ways, such as manually through the Kong Developer portal login psge. However, using the Authentication Action would streamline the user flow since the Action takes the user's full name and the email from the Curity Authenticator and automatically provision that to the Kong Developer Portal using the exposed API.
```

> The [The Kong Developer Portal User Provisioner][curity-kong-dev-portal-user-provisioner] action is available as open-source and can be forked to fit the needs of the environment as needed.

Reviewer 2: ```suggestion
The [The Kong Developer Portal User Provisioner][curity-kong-dev-portal-user-provisioner] action is available as open source and can be forked to fit the needs of the environment as needed.
```

#### Configuration

This Action is straightforward to configure. An HTTP Client is needed to communicate with the Kong Dev Portal API. By default, the **HTTP Authentication** can be left out. Only a correct scheme needs to be configured (HTTP or HTTPS).

The Action also configures the URL to the registration endpoint of the Kong Developer Portal. Here the scheme needs to match what's configured in the HTTP Client used.

![Kong Dev Portal User Provisioner](/assets/images/docs/dev-portal/curity/kong-dev-portal-action.png)

> When the action is created, it can be assigned to the Authenticator(s) used in the client configured in the Curity Identity Server as described above.

Reviewer 2: ```suggestion
When the action is created, it can be assigned to the Authenticators used in the client configured in the Curity Identity Server as described above.
```

#### Action to Resolve Additional User Information

Depending on the Authenticator used, an additional Action may be needed to resolve additional information. By default, The Kong Developer portal provisioning requires `Full Name` and `email`. If the Authenticator does not provide this, it's possible to use an Action to resolve the data. This could be as simple as a **Data Source** action configured to use a Data Source that provides the information.

![Chain Actions](/assets/images/docs/dev-portal/curity/authentication-and-actions.png)

> By default, the Kong Developer Portal Provisioner Action works on the default account table schema of the Curity Identity Server database. This provides `email` as a column, but the `Full Name` is not readily available. The Action will operate on the `attributes` column and parse the information to pass the user's Full Name to the Kong Dev Portal.

Reviewer 2: ```suggestion
By default, the Kong Developer Portal Provisioner Action works on the default account table schema of the Curity Identity Server database. This provides `email` as a column, but the `Full Name` is not readily available. The Action operates on the `attributes` column and parse the information to pass the user's Full Name to the Kong Dev Portal.
```

> The attributes column will contain this structure:

Reviewer 2: ```suggestion
The attributes column contains this structure:
```

```json
{"emails":[{"value":"alice@example.com","primary":true}],"phoneNumbers":[{"value":"555-123-1234","primary":true}],"name":{"givenName":"alice","familyName":"anderson"},"agreeToTerms":"on","urn:se:curity:scim:2.0:Devices":[]}
```

The data source used to resolve additional information needs to be configured with an appropriate **Attribute Query**. This would look similar to this:

```sql
select * from "accounts" where "username"= :subject
```

### Conclusion

> With relatively simple configurations in both the Curity Identity Server and the Kong Developer Portal, it's possible to leverage Curity as the Identity Provider for the Kong Portal. This will provide a very seamless flow for user authentication to the Kong Dev Portal. With the added capability of an Authentication Action, it is possible to automatically provision the user to the Kong Dev Portal for an even more streamlined experience.

Reviewer 2: ```suggestion
With relatively simple configurations in both the Curity Identity Server and the Kong Developer Portal, it's possible to leverage Curity as the Identity Provider for the Kong Dev Portal. This provides a very seamless flow for user authentication to the Kong Dev Portal. With the added capability of an Authentication Action, it is possible to automatically provision the user to the Kong Dev Portal for an even more streamlined experience.
```

[kong-add-service]: /enterprise/{{page.kong_version}}/kong-manager/add-service
[curity-phantom-token-introspection]: https://curity.io/resources/learn/introspect-with-phantom-token
[curity-getting-started]: https://curity.io/resources/getting-started
[curity-phantom-token-pattern]: https://curity.io/resources/learn/phantom-token-pattern
[curity-code-flow-tutorial]: https://curity.io/resources/learn/code-flow
[curity-kong-dev-portal-user-provisioner]: https://curity.io/resources/learn/provision-kong-dev-portal-user
[kong-dev-portal-doc]: https://docs.konghq.com/enterprise/latest/developer-portal

> [kong-dev-portal-doc-oidc]: https://docs.konghq.com/enterprise/2.3.x/developer-portal/configuration/authentication/oidc

Reviewer 2: ```suggestion
[kong-dev-portal-doc-oidc]: https://docs.konghq.com/enterprise/latest/developer-portal/configuration/authentication/oidc
```
